File: 52 Kinema/python/kinema.py
"""
KINEMA

A kinematics physics quiz.

Ported by Dave LeCompte
"""

import random

# We approximate gravity from 9.8 meters/second squared to 10, which
# is only off by about 2%. 10 is also a lot easier for people to use
# for mental math.

g = 10

# We only expect the student to get within this percentage of the
# correct answer. This isn't rocket science.

EXPECTED_ACCURACY_PERCENT = 15


def print_with_tab(spaces_count, msg):
    if spaces_count > 0:
        spaces = " " * spaces_count
    else:
        spaces = ""
    print(spaces + msg)


def do_quiz():
    print()
    print()
    num_questions_correct = 0

    # pick random initial velocity
    v0 = random.randint(5, 40)
    print(f"A BALL IS THROWN UPWARDS AT {v0} METERS PER SECOND.")
    print()

    answer = v0 ** 2 / (2 * g)
    num_questions_correct += ask_player("HOW HIGH WILL IT GO (IN METERS)?", answer)

    answer = 2 * v0 / g
    num_questions_correct += ask_player(
        "HOW LONG UNTIL IT RETURNS (IN SECONDS)?", answer
    )

    t = 1 + random.randint(0, 2 * v0) // g
    answer = v0 - g * t
    num_questions_correct += ask_player(
        f"WHAT WILL ITS VELOCITY BE AFTER {t} SECONDS?", answer
    )

    print()
    print(f"{num_questions_correct} right out of 3.")
    if num_questions_correct >= 2:
        print("  NOT BAD.")


def ask_player(question, answer):
    print(question)
    player_answer = float(input())

    accuracy_frac = EXPECTED_ACCURACY_PERCENT / 100.0
    if abs((player_answer - answer) / answer) < accuracy_frac:
        print("CLOSE ENOUGH.")
        score = 1
    else:
        print("NOT EVEN CLOSE....")
        score = 0
    print(f"CORRECT ANSWER IS {answer}")
    print()
    return score


def main():
    print_with_tab(33, "KINEMA")
    print_with_tab(15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")
    print()
    print()
    print()

    while True:
        do_quiz()


if __name__ == "__main__":
    main()

File: 52 Kinema/perl/kinema.pl
#!/usr/bin/perl
use strict;


print ' 'x 33 . "KINEMA\n";
print ' 'x 15 . "CREATIVE COMPUTING MORRISTOWN, NEW JERSEY\n";
print "\n\n\n";

while (1) {
	print "\n";
	print "\n";
	my $Q=0;
	my $V=5+int(35*rand(1));
	print "A BALL IS THROWN UPWARDS AT $V METERS PER SECOND.\n";
	print "\n";

	my $A=.05*$V^2;
	print "HOW HIGH WILL IT GO (IN METERS)";
	$Q+= &Input($A);

	$A=$V/5;
	print "HOW LONG UNTIL IT RETURNS (IN SECONDS)";
	$Q+= &Input($A);

	my $T=1+int(2*$V*rand(1))/10;
	$A=$V-10*$T;
	print "WHAT WILL ITS VELOCITY BE AFTER $T SECONDS";
	$Q+= &Input($A);

	print "\n";
	print "$Q RIGHT OUT OF 3.";
	if ($Q<2) { next; }
	print " NOT BAD.\n";
	}

exit;


#Line500:
sub Input {
	my ($A)= @_;
	my $Point=0;
	print "? "; chomp(my $G = <STDIN>);
	if (abs(($G-$A)/$A)<.15) {
		print "CLOSE ENOUGH.\n";
		$Point=1;
		} else {
		print "NOT EVEN CLOSE....\n";
		}
	print "CORRECT ANSWER IS $A\n";
	print "\n";
	return $Point;
	}
